src/control/lib/control/pool.go

```diff
@@ -1279,27 +1293,38 @@ func GetMaxPoolSize(ctx context.Context, log logging.Logger, rpcClient UnaryInvo
 					continue
 				}
 
+				if len(ranks) != 0 && !smdDevice.Rank.InList(ranks) {
+					log.Debugf("Skipping SMD device %s (instance %d, ctrlr %s, rank %s) not in ranklist: %s",
+						smdDevice.UUID, smdDevice.Rank, smdDevice.TrAddr, smdDevice.Rank, ranks)
+					continue
+				}
+
 				nvmeRanksBytes[smdDevice.Rank] += smdDevice.AvailBytes
-				log.Debugf("Adding SMD device %s (instance %d, ctrlr %s) is usable: "+
-					"device state=%q, size=%d total=%d",
+				log.Debugf("Adding SMD device %s (instance %d, ctrlr %s) is usable: device state=%q, size=%d total=%d",
 					smdDevice.UUID, smdDevice.Rank, smdDevice.TrAddr,
 					smdDevice.NvmeState.String(), smdDevice.AvailBytes,
 					nvmeRanksBytes[smdDevice.Rank])
 			}
 		}
-		for _, nvmeRankBytes := range nvmeRanksBytes {
+		for rank, nvmeRankBytes := range nvmeRanksBytes {
 			if nvmeBytes > nvmeRankBytes {
 				nvmeBytes = nvmeRankBytes
 			}
+			if _, ok := ranksWithoutSmd[rank]; ok {
```

Comment:
Currently the logic is a little convoluted and probably prone to bugs as a result. I think there is scope for simplifying and breaking out the logic to improve clarity. Specifically, adding a map entry for each rank then deleting those with NVMe space available seems a little clunky, would it be possible to break out of a loop with nvmeBytes 0 if rank is found to have no nvme available?

Reply:
If I understand correctly, you propose to split the nested loops in two:
1. one for computing the minimal size (as it was done previously).
2. one for checking if all usable ranks have smd devices.
For me this solution have the following pros and cons:
+ clearly separate the two process and thus make the code easier to read
- less efficient as we will have to do the same iteration two times
- more code duplication as we will have to do more or less the same tests for checking if the devices are usable or not.
From my side, the efficiency should not be a real issue (however I have no idea of the maximal number of ranks that we could have to process), however the code duplication is more a concern for me.

If we do not want to use a map as a marking board, the only solution that I imagine should be to use two rank sets and checks the subset of their intersection. FMHO, this solution will probably be not more efficient nor readable. However, you have maybe some better solution in mind ?